### PR TITLE
[RISCV][GISel] Use CCValAssign::getCustomReg for converting f16/f32<->GPR.

### PR DESCRIPTION
**Pull request by topperc**
This gives us much better control of the generated code for GISel. I've tried to closely match the current gisel code, but it looks like we had 2 layers of G_ANYEXT in some cases before.

SelectionDAG now checks needsCustom() instead of detecting the special cases in the Bitcast handler.

Unfortunately, IRTranslator for bitcast still generates copies between register classes of different sizes. Because of this we can't handle i16<->f16 bitcasts without crashing. Not sure if I should teach RISCVInstrInfo::copyPhysReg to allow copies between FPR16 and GPR or if I should convert the copies to instructions in GISel.

### CODE REVIEW

**Comment by arsenm** [comment]
> Unfortunately, IRTranslator for bitcast still generates copies between register classes of different sizes. Because of this we can't handle i16<->f16 bitcasts without crashing. Not sure if I should teach RISCVInstrInfo::copyPhysReg to allow copies between FPR16 and GPR or if I should convert the copies to instructions in GISel.

What kind of code is required to emit such a copy? I'd assume that could be directly handled in copyPhysReg.

As as a side note, it's deeply annoying that the CC lowering code operates on legalized types and not the raw IR type

**Comment by topperc** [comment]
> > Unfortunately, IRTranslator for bitcast still generates copies between register classes of different sizes. Because of this we can't handle i16<->f16 bitcasts without crashing. Not sure if I should teach RISCVInstrInfo::copyPhysReg to allow copies between FPR16 and GPR or if I should convert the copies to instructions in GISel.
> 
> What kind of code is required to emit such a copy? I'd assume that could be directly handled in copyPhysReg.

It's just one instruction. I've just never found much evidence that other targets have mismatched copy sizes in copyPhysReg.

**Comment by topperc** [comment]
I've remove the new G_ instructions for GISel and tried to match the original code. I feel like Custom is the right way to describe this for SelectionDAG.

There's some additional cleanup that can be done in RISCVISelLowering, but I think its easier to review without adding more changes.

File: llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp [hunk 1]
```diff
@@ -246,16 +256,29 @@ struct RISCVIncomingValueHandler : public CallLowering::IncomingValueHandler {
   unsigned assignCustomValue(CallLowering::ArgInfo &Arg,
                              ArrayRef<CCValAssign> VAs,
                              std::function<void()> *Thunk) override {
+    const CCValAssign &VA = VAs[0];
+    if ((VA.getLocVT() == MVT::i64 && VA.getValVT() == MVT::f32) ||
+        (VA.getLocVT().isInteger() && VA.getValVT() == MVT::f16)) {
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:261]: maybe consider bfloat but I guess that's totally busted in globalisel anyway

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:261]: I'll do it as a separate patch. I don't think it belongs in this one.

**Comment by topperc** [comment]
Ping

**Comment by topperc** [comment]
Ping

**Review by arsenm — APPROVED** [review]

**Pull request merged**